### ISSUE TITLE
Can not reproduce SciERC results reported in paper

### ISSUE THREAD

**Issue by knuser** [dwadden/dygiepp#3]
Hello,

Thanks for great work and paper!

I'm trying to reproduce SciERC results, but I'm getting only:
```
"best_validation__rel_f1": 0.33248730964467005,
"best_validation_ner_f1": 0.6711409395972655,
```
Paper is reporting 48.4 F1 score on relations extraction on SciERC.

My questions are:
1. Are there any additional steps to take, besides downloading SciERC dataset, to train model and reproduce F1 scores?
1. I have noticed that that validation loss is (almost constantly) increasing every epoch (train loss is dropping). Is this by expected behavior?
1. What was the reason that trainer decided to stop train script earlier with message "INFO - allennlp.training.trainer - Ran out of patience.  Stopping training."?

Bellow I'm placing full installation step-by-step, log tail, system info and pip list:

Installation:
```
conda create --name dygie python=3.7
conda activate dygie
conda install pytorch==1.2.0 torchvision==0.4.0 cudatoolkit=10.0 -c pytorch
pip install allennlp
pip install botocore
git clone https://github.com/dwadden/dygiepp.git
cd dygiepp/
pip install -r requirements.txt
bash ./scripts/data/get_scierc.sh 
bash ./scripts/train/train_scierc.sh 0
```

Log tail:
```
[...]
2019-11-13 16:08:07,639 - WARNING - root - NaN or Inf found in input tensor.                                                                                                                                                                                                       
2019-11-13 16:08:07,640 - WARNING - root - NaN or Inf found in input tensor.                                                                                                                                                                                                       
ner_precision: 0.9754, ner_recall: 0.9725, ner_f1: 0.9740, loss: 4.5208 ||: 100%|#########9| 499/500 [01:05<00:00,  8.05it/s]2019-11-13 16:08:20,732 - WARNING - root - NaN or Inf found in input tensor.                                                                          
2019-11-13 16:08:20,732 - WARNING - root - NaN or Inf found in input tensor.                                                                                                                                                                                                       
2019-11-13 16:08:20,734 - WARNING - root - NaN or Inf found in input tensor.                                                                                                                                                                                                       
2019-11-13 16:08:20,734 - WARNING - root - NaN or Inf found in input tensor.                                                                                                                                                                                                       
2019-11-13 16:08:20,740 - WARNING - root - NaN or Inf found in input tensor.                                                                                                                                                                                                       
2019-11-13 16:08:20,740 - WARNING - root - NaN or Inf found in input tensor.                                                                                                                                                                                                       
2019-11-13 16:08:20,742 - WARNING - root - NaN or Inf found in input tensor.                                                                                                                                                                                                       
2019-11-13 16:08:20,742 - WARNING - root - NaN or Inf found in input tensor.                                                                                                                                                                                                       
ner_precision: 0.9755, ner_recall: 0.9726, ner_f1: 0.9740, loss: 4.5118 ||: 100%|##########| 500/500 [01:05<00:00,  7.62it/s]                                                                                                                                                      
2019-11-13 16:08:20,746 - INFO - allennlp.training.trainer - Validating                                                                                                                                                                                                            
ner_precision: 0.6449, ner_recall: 0.6873, ner_f1: 0.6655, loss: 232.1718 ||: 100%|##########| 50/50 [00:02<00:00, 21.36it/s]                                                                                                                                                      
2019-11-13 16:08:23,090 - INFO - allennlp.training.trainer - Ran out of patience.  Stopping training.                                                                                                                                                                              
2019-11-13 16:08:23,090 - INFO - allennlp.training.checkpointer - loading best weights                                                                                                                                                                                             
2019-11-13 16:08:23,219 - INFO - allennlp.commands.train - To evaluate on the test set after training, pass the 'evaluate_on_test' flag, or use the 'allennlp evaluate' command.                                                                                                   
2019-11-13 16:08:23,219 - INFO - allennlp.models.archival - archiving weights and vocabulary to ./models/scierc/model.tar.gz                                                                                                                                                       
2019-11-13 16:08:40,237 - INFO - allennlp.common.util - Metrics: {
  "best_epoch": 23,                                                                                                                                                                                                                                                                
  "peak_cpu_memory_MB": 3783.14,                                                                                                                                                                                                                                                   
  "peak_gpu_0_memory_MB": 8654,                                                                                                                                                                                                                                                    
  "training_duration": "0:43:55.006665",                                                                                                                                                                                                                                           
  "training_start_epoch": 0,                                                                                                                                                                                                                                                       
  "training_epochs": 37,                                                                                                                                                                                                                                                           
  "epoch": 37,                                                                                                                                                                                                                                                                     
  "training__coref_precision": 0.8032079153669152,                                                                                                                                                                                                                                 
  "training__coref_recall": 0.52068901932887,                                                                                                                                                                                                                                      
  "training__coref_f1": 0.6030049906804981,                                                                                                                                                                                                                                        
  "training__coref_mention_recall": 0.9909547738693467,                                                                                                                                                                                                                            
  "training_ner_precision": 0.9815991970558715,                                                                                                                                                                                                                                    
  "training_ner_recall": 0.9786524349566378,                                                                                                                                                                                                                                       
  "training_ner_f1": 0.9801236011357441,
  "training__rel_precision": 0.8399269628727937,
  "training__rel_recall": 0.8051341890315052,
  "training__rel_f1": 0.8221626452189454,
  "training__rel_span_recall": 0.837222870478413,
  "training__trig_id_precision": 0,
  "training__trig_id_recall": 0,
  "training__trig_id_f1": 0,
  "training__trig_class_precision": 0,
  "training__trig_class_recall": 0,
  "training__trig_class_f1": 0,
  "training__arg_id_precision": 0,
  "training__arg_id_recall": 0,
  "training__arg_id_f1": 0,
  "training__arg_class_precision": 0,
  "training__arg_class_recall": 0,
  "training__arg_class_f1": 0,
  "training__args_multiple": 0,
  "training_loss": 4.156547112312106,
  "training_cpu_memory_MB": 3783.14,
  "training_gpu_0_memory_MB": 8654,
  "validation__coref_precision": 0.5785864145810381,
  "validation__coref_recall": 0.40000249926139114,
  "validation__coref_f1": 0.47185963848741963,
  "validation__coref_mention_recall": 0.9338235294117647,
  "validation_ner_precision": 0.6489988221436984,
  "validation_ner_recall": 0.6836228287841191,
  "validation_ner_f1": 0.6658610271902824,
  "validation__rel_precision": 0.3944954128440367,
  "validation__rel_recall": 0.378021978021978,
  "validation__rel_f1": 0.3860830527497194,
  "validation__rel_span_recall": 0.4175824175824176,
  "validation__trig_id_precision": 0,
  "validation__trig_id_recall": 0,
  "validation__trig_id_f1": 0,
  "validation__trig_class_precision": 0,
  "validation__trig_class_recall": 0,
  "validation__trig_class_f1": 0,
  "validation__arg_id_precision": 0,
  "validation__arg_id_recall": 0,
  "validation__arg_id_f1": 0,
  "validation__arg_class_precision": 0,
  "validation__arg_class_recall": 0,
  "validation__arg_class_f1": 0,
  "validation__args_multiple": 0,
  "validation_loss": 223.13119384765625,
  "best_validation__coref_precision": 0.5206815500675314,
  "best_validation__coref_recall": 0.39612618358360496,
  "best_validation__coref_f1": 0.44944354440112394,
  "best_validation__coref_mention_recall": 0.9375,
  "best_validation_ner_precision": 0.6602641056422568,
  "best_validation_ner_recall": 0.6823821339950371,
  "best_validation_ner_f1": 0.6711409395972655,
  "best_validation__rel_precision": 0.3933933933933934,
  "best_validation__rel_recall": 0.2879120879120879,
  "best_validation__rel_f1": 0.33248730964467005,
  "best_validation__rel_span_recall": 0.3208791208791209,
  "best_validation__trig_id_precision": 0,
  "best_validation__trig_id_recall": 0,
  "best_validation__trig_id_f1": 0,
  "best_validation__trig_class_precision": 0,
  "best_validation__trig_class_recall": 0,
  "best_validation__trig_class_f1": 0,
  "best_validation__arg_id_precision": 0,
  "best_validation__arg_id_recall": 0,
  "best_validation__arg_id_f1": 0,
  "best_validation__arg_class_precision": 0,
  "best_validation__arg_class_recall": 0,
  "best_validation__arg_class_f1": 0,
  "best_validation__args_multiple": 0,
  "best_validation_loss": 142.15479759216308
}
```

System:
```
Ubuntu 18.04.3 LTS
GPU 0: GeForce RTX 2080 Ti
Driver Version: 418.88
CUDA Version: 10.1
```

pip list:
```
$ pip list                                                                                                                                                                                                                                                                   
Package                       Version                                                                                                                                                                                                                                              
----------------------------- -------------------                                                                                                                                                                                                                                  
alabaster                     0.7.12                                                                                                                                                                                                                                               
allennlp                      0.9.0                                                                                                                                                                                                                                                
atomicwrites                  1.3.0                                                                                                                                                                                                                                                
attrs                         19.3.0                                                                                                                                                                                                                                               
Babel                         2.7.0                                                                                                                                                                                                                                                
beautifulsoup4                4.8.1                                                                                                                                                                                                                                                
blis                          0.2.4                                                                                                                                                                                                                                                
boto3                         1.10.16                                                                                                                                                                                                                                              
botocore                      1.13.16                                                                                                                                                                                                                                              
certifi                       2019.9.11                                                                                                                                                                                                                                            
cffi                          1.13.1                                                                                                                                                                                                                                               
chardet                       3.0.4                                                                                                                                                                                                                                                
Click                         7.0                                                                                                                                                                                                                                                  
conllu                        1.3.1                                                                                                                                                                                                                                                
cycler                        0.10.0                                                                                                                                                                                                                                               
cymem                         2.0.2                                                                                                                                                                                                                                                
docutils                      0.15.2                                                                                                                                                                                                                                               
editdistance                  0.5.3                                                                                                                                                                                                                                                
flaky                         3.6.1                                                                                                                                                                                                                                                
Flask                         1.1.1                                                                                                                                                                                                                                                
Flask-Cors                    3.0.8                                                                                                                                                                                                                                                
ftfy                          5.6                                                                                                                                                                                                                                                  
gevent                        1.4.0                                                                                                                                                                                                                                                
greenlet                      0.4.15                                                                                                                                                                                                                                               
h5py                          2.10.0                                                                                                                                                                                                                                               
idna                          2.8                                                                                                                                                                                                                                                  
imagesize                     1.1.0                                                                                                                                                                                                                                                
importlib-metadata            0.23                                                                                                                                                                                                                                                 
itsdangerous                  1.1.0                                                                                                                                                                                                                                                
Jinja2                        2.10.3
jmespath                      0.9.4
joblib                        0.14.0
jsonnet                       0.14.0
jsonpickle                    1.2
kiwisolver                    1.1.0
lxml                          4.4.1
MarkupSafe                    1.1.1
matplotlib                    3.1.1
mkl-fft                       1.0.15
mkl-random                    1.1.0
mkl-service                   2.3.0
more-itertools                7.2.0
murmurhash                    1.0.2
nltk                          3.4.5
numpy                         1.17.3
numpydoc                      0.9.1
olefile                       0.46
overrides                     2.5
packaging                     19.2
pandas                        0.25.3
parsimonious                  0.8.1
Pillow                        6.2.1
pip                           19.3.1
plac                          0.9.6
pluggy                        0.13.0
preshed                       2.0.1
protobuf                      3.10.0
py                            1.8.0
pycparser                     2.19
Pygments                      2.4.2
pyparsing                     2.4.5
pytest                        5.2.2
python-dateutil               2.8.0
python-Levenshtein            0.12.0
pytorch-pretrained-bert       0.6.2
pytorch-transformers          1.1.0
pytz                          2019.3
regex                         2019.11.1
requests                      2.22.0
responses                     0.10.6
s3transfer                    0.2.1
scikit-learn                  0.21.3
scipy                         1.3.2
sentencepiece                 0.1.83
setuptools                    41.6.0.post20191030
six                           1.13.0
snowballstemmer               2.0.0
soupsieve                     1.9.5
spacy                         2.1.9
Sphinx                        2.2.1
sphinxcontrib-applehelp       1.0.1
sphinxcontrib-devhelp         1.0.1
sphinxcontrib-htmlhelp        1.0.2
sphinxcontrib-jsmath          1.0.1
sphinxcontrib-qthelp          1.0.2
sphinxcontrib-serializinghtml 1.1.3
sqlparse                      0.3.0
srsly                         0.2.0
tensorboardX                  1.9
thinc                         7.0.8
torch                         1.2.0
torchvision                   0.4.0a0+6b959ee
tqdm                          4.38.0
Unidecode                     1.1.1
urllib3                       1.25.7
wasabi                        0.4.0
wcwidth                       0.1.7
Werkzeug                      0.16.0
wheel                         0.33.6
word2number                   1.1
zipp                          0.6.0
```

**Comment by ardellelee** [comment]
I got similar results as @knuser , which are significantly lower than the paper reported on SciERC dataset:
`"best_validation__rel_f1": 0.39096045197740115, 
"best_validation_ner_f1": 0.6819277108433234`
The performance is achived at the 30th epoch. The training process is early-stopped after another 15 epochs.

Very curious if `scierc_working_example.jsonnet` is the same config as used in paper. Is it possible to provide some clues for us to tune the model?

Btw, any chance to release the preprocessing code for ACE data?


Regards.

**Comment by dwadden** [comment]
Thanks for the feedback. My guess is there's something wrong with the config. I'll take a look today and post here.

I should have the ACE code by next week.

**Comment by dwadden** [comment]
OK, good news (sorta): the config is wrong. I'll update now and run to see what kind of performance I get.

**Comment by dwadden** [comment]
OK, I pushed a fix. First, run `python ./scripts/pretrained/get_scibert.py` to get SciBERT embeddings. Then run `bash ./scripts/train/train_scierc.sh [gpu-id]`. This will train a model and then evaluate on dev and test set. Surprisingly, performance is much better on test.

Here's a snippet of test set performance for me. Relation performance is slightly better than reported in the paper:
```
2019-11-20 11:39:17,561 - INFO - allennlp.commands.evaluate - Finished evaluating.
2019-11-20 11:39:17,562 - INFO - allennlp.commands.evaluate - Metrics:
...
2019-11-20 11:39:17,562 - INFO - allennlp.commands.evaluate - _ner_precision: 0.6765561372891216
2019-11-20 11:39:17,562 - INFO - allennlp.commands.evaluate - _ner_recall: 0.6947431302270012
2019-11-20 11:39:17,562 - INFO - allennlp.commands.evaluate - _ner_f1: 0.6855290303565666
2019-11-20 11:39:17,562 - INFO - allennlp.commands.evaluate - rel_precision: 0.5130830489192264
2019-11-20 11:39:17,562 - INFO - allennlp.commands.evaluate - rel_recall: 0.46303901437371664
2019-11-20 11:39:17,562 - INFO - allennlp.commands.evaluate - rel_f1: 0.4867781975175391
2019-11-20 11:39:17,562 - INFO - allennlp.commands.evaluate - rel_span_recall: 0.49589322381930184
```

Let me know if you can reproduce and I'll close. Thanks for checking on this!

**Comment by knuser** [comment]
Thanks for pushing those fixes!

I have executed training on SciERC three time. First run was without Nvidia-apex installed. Here are my results on test dataset:

```
    "_ner_precision": 0.6678260869565218,
    "_ner_recall": 0.6881720430107527,
    "_ner_f1": 0.677846425419191,
    "rel_precision": 0.4809575625680087,
    "rel_recall": 0.4537987679671458,
    "rel_f1": 0.4669836238774432,
    "rel_span_recall": 0.4928131416837782,
```

```
    "_ner_precision": 0.669151376146789,
    "_ner_recall": 0.6971326164874552,
    "_ner_f1": 0.6828554710356434,
    "rel_precision": 0.4909670563230606,
    "rel_recall": 0.47433264887063653,
    "rel_f1": 0.4825065274151436,
    "rel_span_recall": 0.5215605749486653,
```

```
    "_ner_precision": 0.6772517321016166,
    "_ner_recall": 0.7007168458781362,
    "_ner_f1": 0.6887844979447534,
    "rel_precision": 0.5442092154420921,
    "rel_recall": 0.4486652977412731,
    "rel_f1": 0.4918401800787845,
    "rel_span_recall": 0.48973305954825463,
```

Thank you

**Comment by dwadden** [comment]
OK great, those look comparable to the paper. There are also pretrained models available for download. I'll close this, please post more issues as they arise!
